### ISSUE TITLE
Allow using Version Catalog for dependency excludes

### ISSUE THREAD

**Issue by TWiStErRob** [gradle/gradle#20274]
When using `libs.versions.toml`:
```toml
junit4 = { module = "junit:junit", version= "4.13.2" }
```

### Expected Behavior
The following "just works":
```kotlin
androidTestImplementation(libs.androidx.test.espresso.core) {
	// Exclude JUnit 4, because all tests are using Jupiter.
	exclude(libs.junit4)
	exclude(group = libs.junit4)
	exclude(module = libs.junit4)
}
```

### Current Behavior
I need some polyfills to get the above compiling:
https://github.com/TWiStErRob/net.twisterrob.colorfilters/pull/42/files#diff-7aed6a315ff2314c660f9be9988802354990555ef32724a6e19e321b8cea8d84

(Ignore the fact that these methods are used from Groovy in the PR, the code in this feature request is all Kotlin DSL. There's a different issue for having this experience possible in Groovy: gradle/gradle#2973)

### Context
> How has this issue affected you?

Need to write the same code in every project / publish/maintain/include a library/plugin. Inconvenient to use.

> What are you trying to accomplish?

Remove all hard-coded error-prone strings from build.gradle files using Version Catalogs.

> What other alternatives have you considered?

```kotlin
androidTestImplementation(libs.androidx.test.espresso.core) {
	// Exclude JUnit 4, because all tests are using Jupiter.
	exclude(group = libs.junit4.get().module.group, module = libs.junit4.get().module.name)
}
```

**Comment by melix** [comment]
This is _not_ the intent of version catalogs. We should _not_ encourage use of excludes, rather the opposite. Catalogs are about _agnostic_ library declarations: they don't care about the call site, in the sense that they don't care what they are used for. In particular, what variant they should use: it's about coordinates. IMO such issues should be solved in a different way. In this case it seems that expresso should _not_ have a strong dependency on JUnit4, so it would be better to report the issue upstream, and use component metadata rules instead to fix the problem.

**Comment by zoobert** [comment]
@melix, I don't believe the request is to include a concept of `exclude` within the version catalog. What I read here is the need to replace dependency references with version catalog notation in an existing build.

For example, I currently have this legitimate dependency declaration in my build:
```
implementation("org.jobrunr:jobrunr-spring-boot-starter:5.1.7") {
        exclude("org.springframework", "spring-webmvc")
}
```

Given a valid declaration in the version catalog, I should be able to replace it with this:
```
implementation(libs.jobrunr.springboot) {
        exclude("org.springframework", "spring-webmvc")
}
``` 

Unfortunately, this is not currently possible because the latter declaration incorrectly resolves to the most basic handler within the `DependencyHandler` class, which has no second parameter to handle dependency configuration in the manner that other dependency notation does.

The solution here seems to extend the `DependencyHandler` to correctly identify `ExternalModuleDependencyFactory` as a first-class citizen of the dependency handler (or at least regarding declaring a dependency) rather than relying on the fall-back mechanism. Within Kotlin Gradle, it would simply require this addition:

```
fun DependencyHandler.implementation(
    dependencyNotation: ExternalModuleDependencyFactory, 
    dependencyConfiguration: Action<ExternalModuleDependency>
): ExternalModuleDependency = addDependencyTo(
    this, "implementation", dependencyNotation, dependencyConfiguration
) as ExternalModuleDependency
```

**Comment by TWiStErRob** [comment]
@zoobert what you're saying works for me:
![image](https://user-images.githubusercontent.com/2906988/193633158-f700b37c-e1b6-4a54-a7ef-ec1bc2e0f6d3.png)
it uses this auto-generated accessor in Kotlin DSL:
![image](https://user-images.githubusercontent.com/2906988/193633291-e408f148-9a98-420f-b1a7-26df05e95231.png)

---

But you're right, @melix might've misunderstood my request. It's not about doing anything to version catalogs themselves. Version Catalogs stay agnostic of usage, that's true, but not all Gradle APIs were retrofitted to support version catalog providers. So I'm asking for an overload of `exclude` that supports a `libs` provider so that I can use the **coordinate** I declared in Version Catalogs in an exclusion. In its simplest form:
```kotlin
fun ModuleDependency.exclude(dep: Provider<MinimalExternalModuleDependency>)
```
so that this compiles:
```kotlin
configuration(libs.some.dep) {
    exclude(libs.other.dep)
}
```

---

Side-note about this particular exclude: it's just an example. The community is trying, but there's no progress: 3 year-old issue with tons of votes: https://github.com/android/android-test/issues/224. The junit dependency is necessary because there are classes in the library that are JUnit 4 specific, I'm just not using them, so I exclude the transitive dependency. I do this so it doesn't mess with my imports and classpath, and I can force myself using one and only one library (JUnit Jupiter). Writing a metadata rule (a class with interface, constructor, injection, implementation method, etc.) is way too heavyweight and complex for a simple exclude like this.

**Comment by zoobert** [comment]
@TWiStErRob You're right (as long as you're using a terminal node) 🤦. In my example above, it was resolved to the generic handler because I used a non-terminal node (e.g., `libs.jobrunnr.springboot` vs `libs.jobrunnr.springboot.starter`). Thanks for turning me on to the right answer.

Gradle version catalogs are a great step forward but almost too cutting edge for me to recommend it to teams. The support among IDE vendors is still a bit lacking though it has been over a year since the feature preview. Support for version catalogs exists but is severely crippled without basic navigation or refactoring capabilities. I've not found any IDEs with working alias rename refactoring, declaration navigation, or usage reporting.

**Comment by rocketraman** [comment]
> @zoobert what you're saying works for me:

It does NOT work for me.

```
e: file:///home/raman/source/my-project/build.gradle.kts:139:9: None of the following functions can be called with the arguments supplied: 
public abstract fun <T : Dependency> implementation(dependency: TypeVariable(T), configure: TypeVariable(T).() -> Unit): TypeVariable(T) defined in org.jetbrains.kotlin.gradle.plugin.KotlinDependencyHandler
public open fun <T : Dependency> implementation(dependency: TypeVariable(T), configure: Closure<*>): TypeVariable(T) defined in org.jetbrains.kotlin.gradle.plugin.KotlinDependencyHandler
public open fun implementation(dependencyNotation: String, configure: Closure<*>): ExternalModuleDependency defined in org.jetbrains.kotlin.gradle.plugin.KotlinDependencyHandler
public abstract fun implementation(dependencyNotation: String, configure: ExternalModuleDependency.() -> Unit): ExternalModuleDependency defined in org.jetbrains.kotlin.gradle.plugin.KotlinDependencyHandler
```

**Comment by mlopatkin** [comment]
The current team's decision was shared there:
* https://github.com/gradle/gradle/issues/26367#issuecomment-2120830998
